infra/Pos/Network/Windows/defaultdns.c

```diff
@@ -0,0 +1,75 @@
+#include <winsock2.h>
+#include <iphlpapi.h>
+#include <stdio.h>
+#include <string.h>
+#include <windows.h>
+
+#define MALLOC(x) HeapAlloc(GetProcessHeap(), 0, (x))
+#define FREE(x) HeapFree(GetProcessHeap(), 0, (x))
+
+char *getWindowsDefDnsServer(void) {
+    FIXED_INFO *pFixedInfo;
+    ULONG ulOutBufLen;
+    DWORD dwRetVal;
+
+// This function isn't thread-safe because of this 'static' declaration.
+// Care is required when calling it.
+    static char dnsIPAddress[16] = "";
+    char allocationErr[] = "Error allocating memory needed to call GetNetworkParams\n";
+
+    pFixedInfo = (FIXED_INFO *) MALLOC(sizeof (FIXED_INFO));
+        if (pFixedInfo == NULL) {
+            printf(allocationErr);
+            return dnsIPAddress;
+        }
+        ulOutBufLen = sizeof (FIXED_INFO);
+
+// Make an initial call to GetAdaptersInfo to get the necessary size into the
+// ulOutBufLen variable
+    if (GetNetworkParams(pFixedInfo, &ulOutBufLen) == ERROR_BUFFER_OVERFLOW) {
+        FREE(pFixedInfo);
+        pFixedInfo = (FIXED_INFO *) MALLOC(ulOutBufLen);
+        if (pFixedInfo == NULL) {
+            printf(allocationErr);
+            return dnsIPAddress;
+        }
+    }
+
+    switch (dwRetVal = GetNetworkParams(pFixedInfo, &ulOutBufLen)) {
+    case ERROR_BUFFER_OVERFLOW:
+        FREE(pFixedInfo);
+        pFixedInfo = (FIXED_INFO *) MALLOC(ulOutBufLen);
+        if (pFixedInfo == NULL) {
+            printf(allocationErr);
+        }
+    case ERROR_INVALID_PARAMETER:
+        printf("An invalid parameter was passed to the function\n");
+        break;
+    case ERROR_NO_DATA:
+        printf("No network parameter information exists for the local computer\n");
+        break;
+    case ERROR_NOT_SUPPORTED:
+        printf("The 'GetNetworkParams function' is not supported by the current "
+               "operating system\n");
+        break;
+    case NO_ERROR:
+        strcpy(dnsIPAddress, pFixedInfo->DnsServerList.IpAddress.String);
```

Comment:
The `DnsServerList` field is of type `IP_ADDR_STRING`; if it were `PIP_ADDR_STRING`, which is a pointer to a `IP_ADDR_STRING`, then this would be a problem; since that's not the case, just checking that `GetNetworkParams` returns `NO_ERROR` is enough as this implies that `pFixedInfo` is not `NULL`, which in turn guarantees there'll be at least one DNS server available.

At most, to be on the safer side, we can doublecheck that `pFixedInfo != NULL` once more here.